core/lib/zksync_core/src/metadata_calculator/updater.rs

```diff
@@ -201,10 +203,23 @@ impl TreeUpdater {
                     .witness_generator_dal()
                     .save_witness_inputs(l1_batch_number, object_key, protocol_version_id)
                     .await;
-                storage
+                let mut transaction = storage
+                    .start_transaction()
+                    .await
+                    .expect("could not start DB storage");
+                transaction
+                    .basic_witness_input_producer_dal()
+                    .create_basic_witness_input_producer_job(l1_batch_number)
```

Comment:
Just to understand better: Is it necessary to place job creation in `MetadataCalculator` (as opposed to, say, the state keeper)? Does it depend on the `MetadataCalculator` output? In general, this looks like the distributed orchestration anti-pattern; `MetadataCalculator` orchestrates other components via job queues, which doesn't look like its scope of responsibilities. Could it make sense for the new component to create jobs for itself instead? (By checking new L1 batches without created jobs – seems like a straightforward SQL query.) In this case, it would be closer to choreography pattern (the component reacts to events – encoded as L1 batch entities – emitted by another component, namely the state keeper).

Reply:
you are rasiging a good point. We used to create the witness job in metadata calc because witness job does depend on its output (merkle proofs). Interesting point on creating the jobs "for itself" - I like it. @EmilLuta I'd say feel free to keep as and let's discuss our approach to queues as a whole (would be interesting to see who creates jobs for whom right now)